CountingSort/app_resources/scatter_shader.comp.hlsl

```diff
@@ -0,0 +1,19 @@
+#include "common.hlsl"
+
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+[[vk::push_constant]] PushConstantData pushConstants;
+[[vk::binding(0,0)]] RWStructuredBuffer<uint32_t> scratch;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void main(uint32_t3 ID : SV_GroupThreadID, uint32_t3 GroupID : SV_GroupID)
+{
+    uint32_t index = WorkgroupSize * GroupID.x + ID.x;
+
+    if (index < pushConstants.dataElementCount)
+    {
+        uint32_t value = vk::RawBufferLoad(pushConstants.inputAddress + sizeof(uint32_t) * index);
+        uint32_t address = nbl::hlsl::glsl::atomicAdd(scratch[value], (uint32_t) -1);
+        vk::RawBufferStore(pushConstants.outputAddress + sizeof(uint32_t) * (address - 1), value);
```

Comment:
using an Exclusive Scan might make the code here prettier, less `-1` in the mix.

Btw to optimize, preload `scratch` into `groupshared`, atomics on shared memory are faster.

Reply:
Wouldn't we have to atomicAdd to the vram buffer eventually to keep the histogram up to date?

Reply:
good point, you can "re-perform" the histogramming into `smem` again, then `smem[value] = atomicAdd(scratch[value],smem[value])` to preload

obviously whats faster will depend on the ratio of elements processed per workgroup and the bucket count

Reply:
I didn't understand what you meant by "re-perform" the histogramming

Reply:
re-build the workgroup-local histogram, same way you do in the first compute dispatch before you prefix sum and accumulate globally

Reply:
Done